### DISCUSSION TITLE
Booleans are `int`, positive-only integers are `int`...

### DISCUSSION THREAD

**Discussion by ghost** [libui-ng/libui-ng#145] [General]
In many C projects I notice the representation of non-integral or positive-only values as `int`.

Is there a reason for this pattern? My first thought is backwards-compatibility with ancient compiler standards, but I'm pretty sure `bool` and `unsigned` have been around for a while.

**Comment by cody271** [comment]
> I'm not sure if there exists a name for it, but in many C projects I notice the over-representation of values as `int`. `uiWindowsTextLength` returns `int`, for example, but a quantity of characters cannot reasonably be interpreted as negative...
> 
So usually the purpose of keeping the `int` signed is to allow for `-1` as an "unknown error" result, in cases where always soft-failing to `0` isn't necessarily acceptable. And there is no such function `uiWindowsTextLength`, did you mean `uiWindowsWindowTextWidth/Height`?

> Is there a reason for this pattern? My first thought is backwards-compatibility with ancient compiler standards, but I'm pretty sure `bool` and `unsigned` have been around for a while!
>
Yes exactly, plain C booleans weren't always guaranteed portable between all the various compilers out there in the past.. This specific reason is much less of an issue nowadays, but another reason remains: the [C99 _Bool](https://en.wikipedia.org/wiki/C_data_types#Boolean_type) type is not guaranteed _ABI_ portable! Breaks some FFI based language bindings.

Best solution is to simply provide our own boolean type unfortunately, which has been the plan [for a while](https://github.com/libui-ng/libui-ng/blob/master/ui.h#L50).. Also see [gboolean](https://developer-old.gnome.org/glib/stable/glib-Basic-Types.html#gboolean) on Unix for a good example of this pattern. Its a major breaking change to the API though, so that's why it hasn't happened yet. Hopefully soon!

**Reply by ghost** [reply]
> So usually the purpose of keeping the `int` signed is to allow for `-1` as an "unknown error" result, in cases where always soft-failing to `0` isn't necessarily acceptable.

I can understand that, though (correct me if I'm wrong!) I don't think I've seen `int` used this way in *libui-ng* (or at least in the Windows code, though I would hope that is consistent across OSes). And, `int` is not only used as a return type; `uiWindowSetContentSize`, for example, accepts `int width` and `int height`. I'm not sure if the intention is for `-1` to mean "ignore this argument" in this case, but that is not how `uiWindowSetContentSize` is implemented. Of course, this is just an arbitrary function I have picked. Here's a list of functions that use `int` when maybe they should use `unsigned`:

- `void uiWindowContentSize(uiWindow *w, int *width, int *height);`
- `void uiWindowSetContentSize(uiWindow *w, int width, int height);`
- `uiWindow *uiNewWindow(const char *title, int width, int height, int hasMenubar);`
- `void uiBoxDelete(uiBox *b, int index);`
- `void uiTabInsertAt(uiTab *t, const char *name, int index, uiControl *c);`
- `void uiTabDelete(uiTab *t, int index);`
- `int uiTabNumPages(uiTab *t);`
- `int uiTabMargined(uiTab *t, int index);`
- `void uiTabSetMargined(uiTab *t, int index, int margined);`
- `void uiComboboxInsertAt(uiCombobox *c, int index, const char *text);`
- `void uiComboboxDelete(uiCombobox *c, int index);`
- `int uiComboboxNumItems(uiCombobox *c);`
- `void uiComboboxSetSelected(uiCombobox *c, int index);`
- `void uiAreaSetSize(uiArea *a, int width, int height);`
- `uiArea *uiNewScrollingArea(uiAreaHandler *ah, int width, int height);`
- `int uiFormNumChildren(uiForm *f);`
- `void uiFormDelete(uiForm *f, int index);`
- `void uiGridAppend(uiGrid *g, uiControl *c, int left, int top, int xspan, int yspan, int hexpand, uiAlign halign, int vexpand, uiAlign valign);`
- `void uiGridInsertAt(uiGrid *g, uiControl *c, uiControl *existing, uiAt at, int xspan, int yspan, int hexpand, uiAlign halign, int vexpand, uiAlign valign);`
- `void uiImageAppend(uiImage *i, void *pixels, int pixelWidth, int pixelHeight, int byteStride);`
- `void uiTableModelRowInserted(uiTableModel *m, int newIndex);`
- `void uiTableModelRowChanged(uiTableModel *m, int index);`
- `void uiTableModelRowDeleted(uiTableModel *m, int oldIndex);`
- `void uiTableAppendTextColumn(uiTable *t, const char *name, int textModelColumn, int textEditableModelColumn, uiTableTextColumnOptionalParams *textParams);`
- `void uiTableAppendImageColumn(uiTable *t, const char *name, int imageModelColumn);`
- `void uiTableAppendImageTextColumn(uiTable *t,
	const char *name,
	int imageModelColumn,
	int textModelColumn,
	int textEditableModelColumn,
	uiTableTextColumnOptionalParams *textParams);`
- `void uiTableAppendCheckboxColumn(uiTable *t,
	const char *name,
	int checkboxModelColumn,
	int checkboxEditableModelColumn);`
- `void uiTableAppendCheckboxTextColumn(uiTable *t,
	const char *name,
	int checkboxModelColumn,
	int checkboxEditableModelColumn,
	int textModelColumn,
	int textEditableModelColumn,
	uiTableTextColumnOptionalParams *textParams);`
- `void uiTableAppendProgressBarColumn(uiTable *t,
	const char *name,
	int progressModelColumn);`
- `void uiTableAppendButtonColumn(uiTable *t,
	const char *name,
	int buttonModelColumn,
	int buttonClickableModelColumn);`
- `void uiTableHeaderSetSortIndicator(uiTable *t,
	int column,
	uiSortIndicator indicator);`
- `uiSortIndicator uiTableHeaderSortIndicator(uiTable *t, int column);`
- `int uiTableColumnWidth(uiTable *t, int column);`
- `void uiTableColumnSetWidth(uiTable *t, int column, int width);`

> And there is no such function `uiWindowsTextLength`, did you mean `uiWindowsWindowTextWidth/Height`?

Sorry; I created a `uiWindowTextLength` in my personal copy of the repo and forgot it doesn't exist elsewhere! Though my point could as well apply to `uiWindowsWindowTextWidth/Height`.

> Yes exactly, plain C booleans weren't always guaranteed portable between all the various compilers out there in the past.. This specific reason is much less of an issue nowadays, but another reason remains: the [C99 _Bool](https://en.wikipedia.org/wiki/C_data_types#Boolean_type) type is not guaranteed _ABI_ portable! Breaks some FFI based language bindings.

I thought FFI might be a factor as well, but I didn't realize `_Bool` wasn't guaranteed to be ABI portable. Thanks for letting me know! Makes sense.
 
> Best solution is to simply provide our own boolean type unfortunately, which has been the plan [for a while](https://github.com/libui-ng/libui-ng/blob/master/ui.h#L50).. Also see [gboolean](https://developer-old.gnome.org/glib/stable/glib-Basic-Types.html#gboolean) on Unix for a good example of this pattern. Its a major breaking change to the API though, so that's why it hasn't happened yet. Hopefully soon!

I see!

Thank you for the fast reply, by the way!

**Comment by szanni** [comment]
Another very simple reason is: many of the underlying platform APIs use `int` (even for counts and sizes) and this avoid casting.

There are definitely a few API choices that could be reconsidered. I would say a `bool` type most likely being a priority, as this communicates intent. The whole signed/unsigned business is sadly almost always a mess, I would personally actually like to see `size_t` in many places, but then we run into the problem of `ssize_t` not being in the C standard us not being able to express special values like `-1`.

**Reply by ghost** [reply]
Would a *libui-ng* boolean type just look like this?

```c
typedef int uiBool;
```

**Reply by ghost** [reply]
I would also like to see `size_t`; I chose `unsigned` as a point of discussion as it is the same size as `int`, but yes, `size_t` would be ideal. It would probably suffice in most places. If we need to represent a special value, we could reserve `SIZE_MAX` and make a note in the documentation.

**Reply by cody271** [reply]
> Would a _libui-ng_ boolean type just look like this?
> 
> ```c
> typedef int uiBool;
> ```
Correct. It's a very simple typing fix, but a major breaking change that needs to happen all at once.
